File: Lib/Telerik/Examples/QuickStart/GridView/Filtering/CustomFilterRow/AdvanceFilterControl.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;
using Telerik.WinControls;
using Telerik.WinControls.Data;
using Telerik.WinControls.UI;

namespace Telerik.Examples.WinControls.GridView.Filtering.CustomFilterRow
{
    public partial class AdvanceFilterControl : UserControl
    {
        #region Fields
        
        private RadGridView gridView;
        private FilterDescriptor advancedFilterDescriptor = new FilterDescriptor(String.Empty, FilterOperator.None, null);
        
        #endregion
        
        #region Constructors
        
        public AdvanceFilterControl()
        {
            InitializeComponent();
            WireEvents();
            this.radDdlColumns.SelectedIndexChanged += new Telerik.WinControls.UI.Data.PositionChangedEventHandler(radDdlColumns_SelectedIndexChanged);
        }
        
        public AdvanceFilterControl(RadGridView template) : this()
        {
            this.gridView = template;
        }
        
        #endregion
        
        #region Properties
        
        public RadGridView GridView
        {
            get
            {
                return this.gridView;
            }
            set
            {
                if (this.gridView != value)
                {
                    if (this.gridView != null)
                    {
                        this.gridView.CustomFiltering -= new GridViewCustomFilteringEventHandler(GridView_CustomFiltering);
                    }
                    
                    this.gridView = value;
                    
                    if (this.gridView != null)
                    {
                        this.gridView.CustomFiltering += new GridViewCustomFilteringEventHandler(GridView_CustomFiltering);
                        this.UpdateTheme(this.Controls);
                    }
                    
                    this.InitializeAdvanceView();
                }
            }
        }
        
        private void UpdateTheme(ControlCollection controls)
        {
            foreach (Control currentControl in controls)
            {
                RadControl current = currentControl as RadControl;
                if (current != null)
                {
                    current.ThemeName = this.gridView.ThemeName;
                    RadPageView pageView = current as RadPageView;
                    if (pageView != null)
                    {
                        foreach (RadPageViewPage page in pageView.Pages)
                        {
                            this.UpdateTheme(page.Controls);
                        }
                    }
                    
                    RadDropDownButton dropDown = current as RadDropDownButton;
                    if (dropDown != null)
                    {
                        dropDown.DropDownButtonElement.BorderThickness = new Padding(0);
                    }
                }
            }
        }
        
        #endregion
        
        #region Methods
        
        private void InitializeAdvanceView()
        {
            this.radDdlColumns.DataSource = this.gridView.Columns;
            this.radDdlColumns.ValueMember = "Name";
            this.radDdlColumns.DisplayMember = "Name";
        }
        
        #endregion
        
        #region Event Handlers
        
        private void radDdlColumns_SelectedIndexChanged(object sender, Telerik.WinControls.UI.Data.PositionChangedEventArgs e)
        {
            this.radDropDownButtonOperator.Items.Clear();
            this.rtbValue1.Text = String.Empty;
            this.rtbValue1.Enabled = false;
            
            if (this.radDdlColumns.SelectedItem != null)
            {
                GridViewDataColumn dataColumn = this.radDdlColumns.SelectedItem.DataBoundItem as GridViewDataColumn;
                
                this.radDropDownButtonOperator.Enabled = true;
                this.radDropDownButtonOperator.Text = "No Filter";
                
                this.advancedFilterDescriptor.Operator = FilterOperator.None;
                this.advancedFilterDescriptor.PropertyName = dataColumn.Name;
                
                List<FilterOperationContext> listContext = FilterOperationContext.GetFilterOperations(dataColumn.DataType);
                foreach (FilterOperationContext context in listContext)
                {
                    RadMenuItem menuItem = new RadMenuItem(context.Name, context.Operator);
                    menuItem.Click += new EventHandler(menuItem_Click);
                    this.radDropDownButtonOperator.Items.Add(menuItem);
                }
            }
            else
            {
                this.radDropDownButtonOperator.Enabled = false;
                this.radDropDownButtonOperator.Text = String.Empty;
            }
        }
        
        private void menuItem_Click(object sender, EventArgs e)
        {
            RadMenuItem menuItem = sender as RadMenuItem;
            this.radDropDownButtonOperator.Text = menuItem.Text;
            
            FilterOperator filterOperator = (FilterOperator)menuItem.Tag;
            this.advancedFilterDescriptor.Operator = filterOperator;
            
            this.rtbValue1.Enabled = filterOperator != FilterOperator.None &&
                                     filterOperator != FilterOperator.IsNull &&
                                     filterOperator != FilterOperator.IsNotNull;
        }
        

        
        private void radBtnAdvanceApply_Click(object sender, EventArgs e)
        {
            this.gridView.EndEdit();
            
            GridTableElement tableElement = this.gridView.TableElement;
            MasterGridViewTemplate masterTemplate = this.gridView.MasterTemplate;
            tableElement.BeginUpdate();
            this.gridView.EnableCustomFiltering = false;
            masterTemplate.DataView.Refresh();
            tableElement.EndUpdate(false);
            
            foreach (GridViewRowInfo row in this.gridView.Rows)
            {
                foreach (GridViewCellInfo cell in row.Cells)
                {
                    cell.Style.Reset();
                }
            }
            
            if (this.radDdlColumns.SelectedItem != null)
            {
                object value = null;
                
                if (!String.IsNullOrEmpty(this.rtbValue1.Text))
                {
                    GridViewDataColumn dataColumn = this.radDdlColumns.SelectedItem.DataBoundItem as GridViewDataColumn;
                    
                    if (dataColumn is GridViewDateTimeColumn)
                    {
                        DateTime dateTimeValue;
                        if (DateTime.TryParse(this.rtbValue1.Text, out dateTimeValue))
                        {
                            value = dateTimeValue;
                        }
                        else
                        {
                            RadMessageBox.SetThemeName(this.gridView.ThemeName);
                            RadMessageBox.Show("The entered value is not valid. Please enter valid date", "Error", MessageBoxButtons.OK);
                        }
                    }
                    else if (dataColumn is GridViewDecimalColumn)
                    {
                        decimal decimalValue;
                        if (decimal.TryParse(this.rtbValue1.Text, out decimalValue))
                        {
                            value = decimalValue;
                        }
                        else
                        {
                            RadMessageBox.SetThemeName(this.gridView.ThemeName);
                            RadMessageBox.Show("The entered value is not valid. Please enter valid decimal value", "Error", MessageBoxButtons.OK);
                        }
                    }
                    else
                    {
                        value = RadDataConverter.Instance.Parse(dataColumn, rtbValue1.Text);
                    }
                }
                
                this.advancedFilterDescriptor.Value = value;
            }
            
            this.gridView.FilterDescriptors.BeginUpdate();
            this.gridView.FilterDescriptors.Clear();
            FilterDescriptor descriptor = this.advancedFilterDescriptor.Clone() as FilterDescriptor;
            this.gridView.FilterDescriptors.Add(descriptor);
            this.gridView.FilterDescriptors.EndUpdate();
        }
        
        private void radTextBoxSearch_TextChanged(object sender, EventArgs e)
        {
            this.gridView.EndEdit();
            
            GridTableElement tableElement = this.gridView.TableElement;
            MasterGridViewTemplate masterTemplate = this.gridView.MasterTemplate;
            
            tableElement.BeginUpdate();
            this.gridView.FilterDescriptors.Clear();
            this.gridView.EnableCustomFiltering = true;
            masterTemplate.DataView.Refresh();
            tableElement.EndUpdate(false);
            
            this.gridView.TableElement.ViewElement.UpdateRows(true);
        }
        
        void GridView_CustomFiltering(object sender, GridViewCustomFilteringEventArgs e)
        {
            string searchPattern = this.radTextBoxSearch.Text; 
            
            bool isEmptyPattern = String.IsNullOrEmpty(searchPattern);
            e.Visible = isEmptyPattern;
            
            foreach (GridViewCellInfo cell in e.Row.Cells)
            {
                string value = String.Empty;
                object cellValue = cell.Value;
                
                if (cellValue != null && !Convert.IsDBNull(cellValue))
                {
                    value = Convert.ToString(cellValue);
                }
                
                bool contains = false;
                
                if (!isEmptyPattern)
                {
                    contains = value.Contains(searchPattern);
                    e.Visible |= contains;
                }
                
                gridView.MasterTemplate.SynchronizationService.SuspendEvent(KnownEvents.RowInvalidated);
                if (contains)
                {
                    cell.Style.CustomizeFill = true;
                    cell.Style.BackColor = Color.FromArgb(201, 252, 254);
                }
                else
                {
                    cell.Style.Reset();
                }
                gridView.MasterTemplate.SynchronizationService.ResumeEvent(KnownEvents.RowInvalidated);
            }
        }
        
        #endregion
        
        private void WireEvents()
        {
            this.radTextBoxSearch.TextChanged += new EventHandler(this.radTextBoxSearch_TextChanged);
            this.radBtnAdvanceApply.Click += new EventHandler(this.radBtnAdvanceApply_Click);
           
        }
    }
}

File: Lib/Telerik/Examples/QuickStart/GridView/Filtering/CustomFilterRow/AdvanceFilterControl.vb
Imports Telerik.WinControls
Imports Telerik.WinControls.Data
Imports Telerik.WinControls.UI

Namespace Telerik.Examples.WinControls.GridView.Filtering.CustomFilterRow
	Partial Public Class AdvanceFilterControl
		Inherits UserControl
		#Region "Fields"

'INSTANT VB NOTE: The variable gridView was renamed since Visual Basic does not allow class members with the same name:
		Private gridView_Renamed As RadGridView
		Private advancedFilterDescriptor As New FilterDescriptor(String.Empty, FilterOperator.None, Nothing)

		#End Region

		#Region "Constructors"

		Public Sub New()
            InitializeComponent()
            WireEvents()
			AddHandler radDdlColumns.SelectedIndexChanged, AddressOf radDdlColumns_SelectedIndexChanged

		End Sub

		Public Sub New(ByVal template As RadGridView)
			Me.New()
			Me.gridView_Renamed = template
		End Sub

		#End Region

		#Region "Properties"

		Public Property GridView() As RadGridView
			Get
				Return Me.gridView_Renamed
			End Get
			Set(ByVal value As RadGridView)
				If Me.gridView_Renamed IsNot value Then
					If Me.gridView_Renamed IsNot Nothing Then
						RemoveHandler gridView_Renamed.CustomFiltering, AddressOf GridView_CustomFiltering
					End If

					Me.gridView_Renamed = value

					If Me.gridView_Renamed IsNot Nothing Then
						AddHandler gridView_Renamed.CustomFiltering, AddressOf GridView_CustomFiltering
						Me.UpdateTheme(Me.Controls)
					End If

					Me.InitializeAdvanceView()
				End If
			End Set
		End Property

		Private Sub UpdateTheme(ByVal controls As ControlCollection)
			For Each currentControl As Control In controls
				Dim current As RadControl = TryCast(currentControl, RadControl)
				If current IsNot Nothing Then
					current.ThemeName = Me.gridView_Renamed.ThemeName
                    Dim pageView As RadPageView = TryCast(current, RadPageView)
					If pageView IsNot Nothing Then
						For Each page As RadPageViewPage In pageView.Pages
							Me.UpdateTheme(page.Controls)
						Next page
                    End If
                    Dim dropDown As RadDropDownButton = TryCast(current, RadDropDownButton)
                    If dropDown IsNot Nothing Then
                        dropDown.DropDownButtonElement.BorderThickness = New Padding(0)
                    End If

				End If
			Next currentControl
		End Sub

		#End Region

		#Region "Methods"

		Private Sub InitializeAdvanceView()
			Me.radDdlColumns.DataSource = Me.gridView_Renamed.Columns
			Me.radDdlColumns.ValueMember = "Name"
			Me.radDdlColumns.DisplayMember = "Name"
		End Sub

		#End Region

		#Region "Event Handlers"

		Private Sub radDdlColumns_SelectedIndexChanged(ByVal sender As Object, ByVal e As Telerik.WinControls.UI.Data.PositionChangedEventArgs)
			Me.radDropDownButtonOperator.Items.Clear()
			Me.rtbValue1.Text = String.Empty
			Me.rtbValue1.Enabled = False

			If Me.radDdlColumns.SelectedItem IsNot Nothing Then
				Dim dataColumn As GridViewDataColumn = TryCast(Me.radDdlColumns.SelectedItem.DataBoundItem, GridViewDataColumn)

				Me.radDropDownButtonOperator.Enabled = True
				Me.radDropDownButtonOperator.Text = "No Filter"

				Me.advancedFilterDescriptor.Operator = FilterOperator.None
				Me.advancedFilterDescriptor.PropertyName = dataColumn.Name

				Dim listContext As List(Of FilterOperationContext) = FilterOperationContext.GetFilterOperations(dataColumn.DataType)
				For Each context As FilterOperationContext In listContext
					Dim menuItem As New RadMenuItem(context.Name, context.Operator)
					AddHandler menuItem.Click, AddressOf menuItem_Click
					Me.radDropDownButtonOperator.Items.Add(menuItem)
				Next context
			Else
				Me.radDropDownButtonOperator.Enabled = False
				Me.radDropDownButtonOperator.Text = String.Empty
			End If
		End Sub

		Private Sub menuItem_Click(ByVal sender As Object, ByVal e As EventArgs)
			Dim menuItem As RadMenuItem = TryCast(sender, RadMenuItem)
			Me.radDropDownButtonOperator.Text = menuItem.Text

			Dim filterOperator As FilterOperator = CType(menuItem.Tag, FilterOperator)
			Me.advancedFilterDescriptor.Operator = filterOperator

			Me.rtbValue1.Enabled = filterOperator <> FilterOperator.None AndAlso filterOperator <> FilterOperator.IsNull AndAlso filterOperator <> FilterOperator.IsNotNull
		End Sub

        Private Sub radBtnAdvanceApply_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.gridView.EndEdit()

            Dim tableElement As GridTableElement = Me.gridView.TableElement
            Dim masterTemplate As MasterGridViewTemplate = Me.gridView.MasterTemplate
            tableElement.BeginUpdate()
            Me.gridView.EnableCustomFiltering = False
            masterTemplate.DataView.Refresh()
            tableElement.EndUpdate(False)

            For Each row As GridViewRowInfo In Me.gridView.Rows
                For Each cell As GridViewCellInfo In row.Cells
                    cell.Style.Reset()
                Next cell
            Next row

            If Me.radDdlColumns.SelectedItem IsNot Nothing Then
                Dim value As Object = Nothing

                If Not String.IsNullOrEmpty(Me.rtbValue1.Text) Then
                    Dim dataColumn As GridViewDataColumn = TryCast(Me.radDdlColumns.SelectedItem.DataBoundItem, GridViewDataColumn)

                    If TypeOf dataColumn Is GridViewDateTimeColumn Then
                        Dim dateTimeValue As Date = Nothing
                        If Date.TryParse(Me.rtbValue1.Text, dateTimeValue) Then
                            value = dateTimeValue
                        Else
                            RadMessageBox.SetThemeName(Me.gridView.ThemeName)
                            RadMessageBox.Show("The entered value is not valid. Please enter valid date", "Error", MessageBoxButtons.OK)
                        End If
                    ElseIf TypeOf dataColumn Is GridViewDecimalColumn Then
                        Dim decimalValue As Decimal = Nothing
                        If Decimal.TryParse(Me.rtbValue1.Text, decimalValue) Then
                            value = decimalValue
                        Else
                            RadMessageBox.SetThemeName(Me.gridView.ThemeName)
                            RadMessageBox.Show("The entered value is not valid. Please enter valid decimal value", "Error", MessageBoxButtons.OK)
                        End If
                    Else
                        value = RadDataConverter.Instance.Parse(dataColumn, rtbValue1.Text)
                    End If
                End If

                Me.advancedFilterDescriptor.Value = value
            End If

            Me.gridView_Renamed.FilterDescriptors.BeginUpdate()
            Me.gridView_Renamed.FilterDescriptors.Clear()
            Dim descriptor As FilterDescriptor = TryCast(Me.advancedFilterDescriptor.Clone(), FilterDescriptor)
            Me.gridView_Renamed.FilterDescriptors.Add(descriptor)
            Me.gridView_Renamed.FilterDescriptors.EndUpdate()
        End Sub

		Private Sub radTextBoxSearch_TextChanged(ByVal sender As Object, ByVal e As EventArgs)
			Me.gridView_Renamed.EndEdit()

			Dim tableElement As GridTableElement = Me.gridView_Renamed.TableElement
			Dim masterTemplate As MasterGridViewTemplate = Me.gridView_Renamed.MasterTemplate

			tableElement.BeginUpdate()
			Me.gridView_Renamed.FilterDescriptors.Clear()
			Me.gridView_Renamed.EnableCustomFiltering = True
			masterTemplate.DataView.Refresh()
			tableElement.EndUpdate(False)

			Me.gridView_Renamed.TableElement.ViewElement.UpdateRows(True)
		End Sub

		Private Sub GridView_CustomFiltering(ByVal sender As Object, ByVal e As GridViewCustomFilteringEventArgs)
			Dim searchPattern As String = Me.radTextBoxSearch.Text

			Dim isEmptyPattern As Boolean = String.IsNullOrEmpty(searchPattern)
			e.Visible = isEmptyPattern

			For Each cell As GridViewCellInfo In e.Row.Cells
				Dim value As String = String.Empty
				Dim cellValue As Object = cell.Value

				If cellValue IsNot Nothing AndAlso (Not Convert.IsDBNull(cellValue)) Then
					value = Convert.ToString(cellValue)
				End If

				Dim contains As Boolean = False

				If Not isEmptyPattern Then
					contains = value.Contains(searchPattern)
					e.Visible = e.Visible Or contains
				End If

				gridView_Renamed.MasterTemplate.SynchronizationService.SuspendEvent(KnownEvents.RowInvalidated)
				If contains Then
					cell.Style.CustomizeFill = True
					cell.Style.BackColor = Color.FromArgb(201, 252, 254)
				Else
					cell.Style.Reset()
				End If
				gridView_Renamed.MasterTemplate.SynchronizationService.ResumeEvent(KnownEvents.RowInvalidated)
			Next cell
		End Sub

		#End Region

		Private Sub WireEvents()
			AddHandler radTextBoxSearch.TextChanged, AddressOf radTextBoxSearch_TextChanged
			AddHandler radBtnAdvanceApply.Click, AddressOf radBtnAdvanceApply_Click

		End Sub
	End Class
End Namespace
